### ISSUE TITLE
Missing secret causes the worker to crash

### ISSUE THREAD

**Issue by JordanSussman** [go-vela/community#133]
## Description

Utilizing a missing secret will cause the worker process to crash instead of handling the failure gracefully. You can submit a `.vela.yml` like the one bellow to reproduce the failure:

```yaml
version: "1"

steps:
  - name: test
    image: alpine:latest
    secrets: [ password ]
    commands:
      - env 
    ruleset:
      event: [ pull_request ]
      branch: [ test ] 

secrets:
  - name: password
    engine: vault
    key: foo/bar/star
    type: shared
```

The logs I obtained from the worker are as follows:

```json
{"build":8,"host":"db2942443d93","level":"error","msg":"unable to destroy step: Error: No such container: step_JordanSussman_vela-sample_8_test","repo":"JordanSussman/vela-sample","time":"2020-10-16T14:08:15Z","user":"JordanSussman"}
{"build":8,"host":"db2942443d93","level":"info","msg":"deleting volume","repo":"JordanSussman/vela-sample","time":"2020-10-16T14:08:15Z","user":"JordanSussman"}
{"level":"trace","msg":"removing volume for pipeline JordanSussman_vela-sample_8","time":"2020-10-16T14:08:15Z"}
{"build":8,"host":"db2942443d93","level":"info","msg":"deleting network","repo":"JordanSussman/vela-sample","time":"2020-10-16T14:08:15Z","user":"JordanSussman"}
{"level":"trace","msg":"removing network for pipeline JordanSussman_vela-sample_8","time":"2020-10-16T14:08:15Z"}
{"level":"error","msg":"failing worker executor: unable to pull secrets: unable to retrieve secret: unable to get secret shared/foo/bar/star from vault service: secret secret/data/shared/foo/bar/star does not exist","time":"2020-10-16T14:08:15Z"}
{"level":"error","msg":"failing worker operator: unable to pull secrets: unable to retrieve secret: unable to get secret shared/foo/bar/star from vault service: secret secret/data/shared/foo/bar/star does not exist","time":"2020-10-16T14:08:15Z"}
{"level":"fatal","msg":"shutting down worker","time":"2020-10-16T14:08:15Z"}
```

## Useful Information

1. What is the output of `vela --version`?

0.6.0

2. What operating system is being used?

Linux

3. Any other important details?

**Comment by jbrockopp** [comment]
Conversations on this were had in Slack that explain the draft PR above a bit more in-depth.

cc: @JordanSussman @meinenec Reposting here for visibility:

> Jordan Sussman  9:03 AM
> Hey, any input on https://github.com/go-vela/community/issues/133? I have very mixed opinions on allowing https://github.com/go-vela/worker/blob/0955becfb3f982e44e60b68030e5a592cf22046b/cmd/vela-worker/start.go#L82 to kill the entire worker process instead of just the relevant subprocess.
>
> One could fix that particular bug by adding some error type checking, but I assume the current flow has other common failure scenarios besides a missing secret. Haven’t used the tomb package enough to understand how I would propose changing the current design.
>
> TL;DR the current design of the worker code is not well suited for scenarios where `VELA_BUILD_LIMIT` > 1 since a single build failure can cause the entire process to crash.

> Jordan Brockopp  9:19 AM
> TBH, I actually don't think that's where the issue lies :+1: I say this, because that error handling (IMO) is there for the right reasons.
>
> Essentially what you're seeing there is that we spawn two go routines:
>
> * [one for the server (API)](https://github.com/go-vela/worker/blob/master/cmd/vela-worker/start.go#L37)
> * [one for the operator (build daemon)](https://github.com/go-vela/worker/blob/master/cmd/vela-worker/start.go#L59)
>
> In the event one of those returns an error, then the worker should die.
>
> That being said, I'm not discrediting this as a bug... More so, I'm suggesting that other code needs to be adjusted to fix the bug :+1:
>
> The [operator lives here](https://github.com/go-vela/worker/blob/master/cmd/vela-worker/operate.go#L18). If you haven't reviewed that function, the tl;dr is the operator is responsible for managing executors. At this point, it is our intention that 1 executor will only ever run 1 build. BUT, the worker can have multiple executors running. Each executor is spawned inside a separate errgroup thread as a go routine.
>
> Each executor that is created, [calls the exec function](https://github.com/go-vela/worker/blob/master/cmd/vela-worker/operate.go#L59) which can be found [here](https://github.com/go-vela/worker/blob/master/cmd/vela-worker/exec.go#L22). Inside that function, this is where we actually make calls to the [github.com/go-vela/pkg-executor](http://github.com/go-vela/pkg-executor) codebase. I think the real issue is that when one of the executors experiences an error, we are having them return that error up the stack. I'm wondering if that shouldn't be the case, because like you said, it will cause the other executors to crash and burn.

> Jordan Sussman  9:21 AM
> I have already confirmed that returning nil on https://github.com/go-vela/worker/blob/master/cmd/vela-worker/exec.go#L127 will cause everything to work when the secret is missing.

> Jordan Brockopp  9:22 AM
> I understand the wall of text isn't great, so to summarize, I think we might want to consider **NOT** returning errors on these lines:
> * https://github.com/go-vela/worker/blob/master/cmd/vela-worker/exec.go#L119
> * https://github.com/go-vela/worker/blob/master/cmd/vela-worker/exec.go#L127
> * https://github.com/go-vela/worker/blob/master/cmd/vela-worker/exec.go#L135
> * https://github.com/go-vela/worker/blob/master/cmd/vela-worker/exec.go#L143
>
> In short, the `CreateBuild`, `PlanBuild`, and `AssembleBuild` functions can return errors for multiple reasons... one of them being that it failed to fetch a secret.
>
> Rather than returning that error, I think we should log it and then return nil.
>
> This behavior would cause the executor to crash, BUT, should leave the worker up and running.
>
> Keep in mind that the behavior you are experiencing is that the entire worker is crashing when in reality, we only want that single executor (goroutine) to crash

> Jordan Sussman  9:24 AM
> Wonder if we should instead just return nil on https://github.com/go-vela/worker/blob/master/cmd/vela-worker/operate.go#L66.
>
> ^ would prevent the error from going up the stack.

> Jordan Brockopp  9:27 AM
> I'm also open to that as a suggestion :+1:
>
> TBH, whatever way we decide to do, I think some exhaustive and extensive testing would be required to make sure this doesn't have any unintended consequences.
>
> But that being said, we don't run any of our workers with more than a single executor (we don't have the freedom to have "big boxes" in our cloud services).
>
> So to some extent, we would have to defer some of the real-world testing to you folks :sadpanda:
>
> Doing so locally should be enough to get some code in place for sure though :+1:

> Jordan Sussman  9:28 AM
> I’ll plan to further testing of it this week.
>
> Thanks for the input.
>
> Still thinking through how much executors we want to run on our agents… Likely to still be >= 2.

> Jordan Brockopp  9:35 AM
> Your welcome :slightly_smiling_face: Apologies for the inconvenience!
> And that's likely going to be a growing pain type of scenario.. think it all depends on how much resources you are trying to provide per build type of deal.

> Jordan Sussman  9:36 AM
> Likely to opt for more/smaller agents once the API endpoints to enable autoscaling have been added.

> Emmanuel  11:22 AM
> Sorry I’m a little late to the party :stuck_out_tongue: Brockopp pretty much hit my thoughts as well so :+1:
>
> As was pointed out, there are a lot of growing pains in this area. I believe that a refactor of the executor has been discussed as a possibility for handling all of this better but I think that’s always a given when we run into new use-cases.
